### PR TITLE
Fix contract reminder notifications

### LINKED ISSUE #2596: [9.1.6] - Contract end mail notification checks start date
The system send the mail for contract end date using the start date without considering contract period and renewal period. You can see an example on the screenshot below:

2017-08-10 21_01_18-glpi - contracts - 10

### PR DESCRIPTION
**Pull request by AnaelMobilia**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #2596

  - Fix bad computation for the date of the first alert
  - Disable contracts notification for deleted contracts (trashbin)
  - Simplify coding logic
  - Remove unused vars
Close #2596

### CODE REVIEW

File: inc/contract.class.php [hunk 1]
```diff
@@ -1269,71 +1267,68 @@ static function cronContract($task = null) {
          // Get contrats with periodicity alerts
          $query_periodicity = "SELECT `glpi_contracts`.*
                                FROM `glpi_contracts`
-                               WHERE `glpi_contracts`.`alert` & ".pow(2, Alert::PERIODICITY)." > '0'
-                                     AND `glpi_contracts`.`entities_id` = '".$entity."' ";
+                               WHERE `glpi_contracts`.`alert` & " . pow(2, Alert::PERIODICITY) . " > '0'
```
Comment (trasher) [inline inc/contract.class.php]: Could you please convert this one to use the new querying method?

Comment (trasher) [reply inc/contract.class.php]: In facts, the `&` operator has been added (but is missing in the doc):
```php
$crit = [
   'FROM' => 'glpi_contracts',
   'WHERE' => [
      'alert'        => ['&', pow(2, Alert::PERIODICITY)],
      'entities_id'  => $entity,
      'is_deleted'   => 0
   ]
];
//SELECT * FROM `glpi_contracts` WHERE `alert` & '64' AND `entities_id` = '0' AND `is_deleted` = '0'
```

That said, the `> 0` part cannot be added, but I'm not sure this is required

Comment (AnaelMobilia) [reply inc/contract.class.php]: @trasher After some tests, I agre with you, the > 0 part will not have a real role : 0&0 => false  / 0&1 => false / 1&1 => true

I will update the PR. Thanks !
Note : i made a PR for the docdev

File: inc/contract.class.php [hunk 2]
```diff
@@ -1267,73 +1265,72 @@ static function cronContract($task = null) {
...
-                && ($end_alert > date('Y-m-d'))) {
+            $entity = $data['entities_id'];
+
+            // For contracts with begin date and periodicity
+            if (!empty($data['begin_date']) && $data['periodicity']) {
                $todo = ['periodicity' => Alert::PERIODICITY];
-               if ($data['alert']&pow(2, Alert::NOTICE)) {
-                  $todo['periodicitynotice'] = Alert::NOTICE;
+               if ($data['alert'] & pow(2, Alert::NOTICE)) {
+                  $todo = ['periodicitynotice' => Alert::NOTICE];
```
Comment (trasher) [inline inc/contract.class.php]: Are you sure for this one? This will replace the whole `$todo`.

Comment (AnaelMobilia) [reply inc/contract.class.php]: Nice catch ! Thanks :)

File: inc/contract.class.php [hunk 3]
```diff
@@ -1267,73 +1265,72 @@ static function cronContract($task = null) {
...
-                && ($end_alert > date('Y-m-d'))) {
+            $entity = $data['entities_id'];
+
+            // For contracts with begin date and periodicity
+            if (!empty($data['begin_date']) && $data['periodicity']) {
                $todo = ['periodicity' => Alert::PERIODICITY];
-               if ($data['alert']&pow(2, Alert::NOTICE)) {
-                  $todo['periodicitynotice'] = Alert::NOTICE;
+               if ($data['alert'] & pow(2, Alert::NOTICE)) {
+                  $todo['periodicitynotice'] = Alert::NOTICE];
```
Comment (orthagh) [inline inc/contract.class.php]: trailing `]` after your last commit (tests fails because of this)

**Comment by trasher** [comment]
See #4043 for 9.2 version of this one (due to DBMysqlIterator limitations)

**Pull request merged**